include/sys/spa.h

```diff
@@ -408,6 +488,26 @@ _NOTE(CONSTCOND) } while (0)
 #define	BP_GET_LEVEL(bp)		BF64_GET((bp)->blk_prop, 56, 5)
 #define	BP_SET_LEVEL(bp, x)		BF64_SET((bp)->blk_prop, 56, 5, x)
 
+/* encrypted, authenticated, and MAC cksum bps use the same bit */
+#define	BP_USES_CRYPT(bp)		BF64_GET((bp)->blk_prop, 61, 1)
+#define	BP_SET_CRYPT(bp, x)		BF64_SET((bp)->blk_prop, 61, 1, x)
+
+#define	BP_IS_ENCRYPTED(bp)			\
+	(BP_USES_CRYPT(bp) &&			\
+	BP_GET_LEVEL(bp) <= 0 &&		\
+	DMU_OT_IS_ENCRYPTED(BP_GET_TYPE(bp)))
```

Comment:
> We could use a bit out of BP_GET_TYPE()?

Hmmm. That might be even more work to interpret the bp, depending on what you count as "more work". `DMU_OT_NEW_TYPE` and `DMU_OT_METADATA` are currently using the 2 highest bits: the only bits which aren't used by the normal types in `dmu_object_type_t`.

So the only way we could make that work is if we stole the `0x40` bit which normally means `DMU_OT_METADATA`  to instead mean `DMU_OT_ENCRYPTED` if `DMU_OT_NEWTYPE` is not set, and otherwise `DMU_OT_ENCRYPTED` would be `0x20`, as it currently is. That is a very ugly macro.....

The only 2 other options are to use a new bit from the padding (which isn't very nice because it starts to break into one of our 64 bit words) or to do as you described above (authenticated = encrypted without an IV / salt. I can give that a shot if you would like.